### ISSUE TITLE
Can I get a differentiable voxelization method?

### ISSUE THREAD

**Issue by zyl1336110861** [traveller59/spconv#451]
@traveller59 Just as the title says.  Before voxelization process, I tend to add some conv layers for my research purpose, so I want to get a differentiable voxelization which can pass the gradient flow from the loss to the former conv layers. Thanks very much for your efforts!!!

**Comment by L-Reichardt** [comment]
Just for future reference since I just faced the same issue: Something along the following lines should work (not fully tested or optimized)

```python
        # Dynamic coordinate range per batch min / max
        coords_min = torch.amin(xyz, dim=(0, 1))  # x, y, z
        coords_max = torch.amax(xyz, dim=(0, 1))
        grid_size = torch.ceil(torch.div(coords_max - coords_min, self.voxel_size_xyz))

        # Convert batch to Sparse Tensor
        all_voxels = []
        all_coords = []
        B = xyz.shape[0]  # Batch size

        for batch in range(B):
            # Calculate the coordinate for each voxel
            with torch.no_grad():
                voxel_coords = xyz[batch].clone()
                voxel_coords -= coords_min
                voxel_coords /= self.voxel_size_xyz
                voxel_coords = voxel_coords.int()

                # Keep only the last occuring point as a feature
                unq_voxel_coords, source_ids = torch.unique(
                    voxel_coords, sorted=False, return_inverse=True, dim=0
                )

                # Track which batch each voxel belongs to
                N = unq_voxel_coords.shape[0]
                batch_id = torch.full(
                    (N, 1), batch, dtype=unq_voxel_coords.dtype, device=device
                )
                coords = torch.cat(
                    [batch_id, unq_voxel_coords], dim=1
                )  # (N, [B, x, y, z])

            # Actually voxelize
            gather_ids = torch.unique(source_ids, sorted=False)
            voxel_features = features[batch][gather_ids]
            all_voxels.append(voxel_features)
            all_coords.append(coords)

        # Convert to SparseTensor
        # NOTE: Flip to z,y,x order for indices and shapes
        # https://github.com/traveller59/spconv/blob/master/spconv/pytorch/core.py#L132
        all_coords = torch.cat(all_coords, dim=0)
        all_coords = all_coords[:, [0, 3, 2, 1]]  # flip B,x,y,z -> B,z,y,x
        grid_size = grid_size.flip(0).contiguous()

        sparse_tensor = spconv.SparseConvTensor(
            features=torch.cat(all_voxels, dim=0).squeeze(),  # features
            indices=all_coords,  # z,y,x order
            spatial_shape=grid_size,  # z,y,x order
            batch_size=B,
        )
```
